GoogleContainerTools/skaffold
# fix: add error code for cross-platform build failures

Fixes: #7638 

**Description**

Gives an actionable error message when qemu platform emulators aren't installed and that causes cross-platform build failures

**Testing instructions:**
* Use Cloud Shell to test this feature. Click below button to open this branch on Cloud Shell.
[![Open in Cloud Shell](https://gstatic.com/cloudssh/images/open-btn.svg)](https://ssh.cloud.google.com/cloudshell/editor?cloudshell_git_repo=https://github.com/gsquared94/skaffold&cloudshell_git_branch=fix-7638&cloudshell_open_in_editor=examples/cross-platform-builds/skaffold.yaml)

* Build for `linux/arm64` in `examples/cross-platform-builds`
  ```
  export ROOT=$PWD
  export PROJECT=$ROOT/examples/cross-platform-builds
  
  cd $PROJECT && skaffold build --platform=linux/arm64 --push=false
  ```
* Build should fail with the following error message:
  ```
  => ERROR [builder 5/5] RUN go build -gcflags="${SKAFFOLD_GO_GCFLAGS}" -trimpath -o /app main.go                                                                  1.0s
  ------
  > [builder 5/5] RUN go build -gcflags="${SKAFFOLD_GO_GCFLAGS}" -trimpath -o /app main.go:
  #11 0.714 exec /bin/sh: exec format error
  ------
  executor failed running [/bin/sh -c go build -gcflags="${SKAFFOLD_GO_GCFLAGS}" -trimpath -o /app main.go]: exit code: 1
  exit status 1. Docker build ran into internal error. Please retry.
  If this keeps happening, please open an issue..
  ```
* This error is not actionable and doesn't inform the user how to fix it. Now build `skaffold` from this branch
  ```
  cd $ROOT && LOCAL=true make 
  ```
* Again build for `linux/arm64` in `examples/cross-platform-builds` project with the new `skaffold` binary
  ```
  cd $PROJECT && ../../out/skaffold build --platform=linux/arm64
  ```
* You should now get an actionable error message that explains how to install the missing platform emulators.
  ```
  #11 ERROR: executor failed running [/bin/sh -c go build -gcflags="${SKAFFOLD_GO_GCFLAGS}" -trimpath -o /app main.go]: exit code: 1
  ------
   > [builder 5/5] RUN go build -gcflags="${SKAFFOLD_GO_GCFLAGS}" -trimpath -o /app main.go:
  ------
  executor failed running [/bin/sh -c go build -gcflags="${SKAFFOLD_GO_GCFLAGS}" -trimpath -o /app main.go]: exit code: 1
  running build: exit status 1. To run cross-platform builds, check that QEMU platform emulators are installed correctly. To install, run:
  
          docker run --privileged --rm tonistiigi/binfmt --install all
  
  For more details, see https://registry.hub.docker.com/r/tonistiigi/binfmt.
  ```

* Now, execute the command specified in the actionable error message. This will install the required platform emulators.
  ```
  docker run --privileged --rm tonistiigi/binfmt --install all
  ```
* Again build for `linux/arm64` in `examples/cross-platform-builds` project with the new `skaffold` binary. The build should now succeed
  ```
  cd $PROJECT && ../../out/skaffold build --platform=linux/arm64
  ```
  ```
  Starting build...
  Building [skaffold-example]...
  ...
  #13 exporting to image
  #13 sha256:e8c613e07b0b7ff33893b694f7759a10d42e180f2b4dc349fb57dc6b71dcab00
  #13 exporting layers done
  #13 writing image sha256:ef8330aab73bfd0df02caf7c7a3235100c6345b4c57192cdefe4a50a5b0f119c
  ```
